File: src/main/java/cc/mirukuneko/realtrainmodrenewed/blockentity/BallastBlockEntity.java
package cc.mirukuneko.realtrainmodrenewed.blockentity;

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewedBlockEntities;
import net.minecraft.core.BlockPos;
import net.minecraft.core.HolderLookup;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket;
import net.minecraft.world.level.storage.ValueInput;
import net.minecraft.world.level.storage.ValueOutput;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;

/**
 * 道床(バラスト)ブロックのブロックエンティティ。
 * 対応するレールコア(LargeRailCore)の位置を保持し、道床を壊すとレールも撤去できるようにする。
 * また列車設置時のレール検出 (getRailMapAt) で道床からレールコアを引けるようにする。
 */
public class BallastBlockEntity extends BlockEntity {
    private BlockPos corePos;

    public BallastBlockEntity(BlockPos pos, BlockState state) {
        super(RealTrainModRenewedBlockEntities.BALLAST.get(), pos, state);
    }

    public void setCorePos(BlockPos corePos) {
        this.corePos = corePos;
        setChanged();
    }

    public BlockPos getCorePos() {
        return corePos;
    }

    @Override
    protected void saveAdditional(ValueOutput tag) {
        super.saveAdditional(tag);
        if (corePos != null) {
            tag.putIntArray("CorePos", new int[]{corePos.getX(), corePos.getY(), corePos.getZ()});
        }
    }

    @Override
    protected void loadAdditional(ValueInput tag) {
        super.loadAdditional(tag);
        tag.getIntArray("CorePos").ifPresent(a -> {
            if (a.length >= 3) {
                corePos = new BlockPos(a[0], a[1], a[2]);
            }
        });
    }

    @Override
    public CompoundTag getUpdateTag(HolderLookup.Provider registries) {
        return this.saveWithoutMetadata(registries);
    }

    @Override
    public ClientboundBlockEntityDataPacket getUpdatePacket() {
        return ClientboundBlockEntityDataPacket.create(this);
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/blockentity/BallastBlockEntity.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.blockentity

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewedBlockEntities
import net.minecraft.core.BlockPos
import net.minecraft.core.HolderLookup
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket
import net.minecraft.world.level.storage.ValueInput
import net.minecraft.world.level.storage.ValueOutput
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.state.BlockState

class BallastBlockEntity(pos: BlockPos, state: BlockState) :
    BlockEntity(RealTrainModRenewedBlockEntities.BALLAST.get(), pos, state) {

    var corePos: BlockPos? = null
        private set

    fun setCorePos(corePos: BlockPos?) { this.corePos = corePos; setChanged() }

    override fun saveAdditional(tag: ValueOutput) {
        super.saveAdditional(tag)
        corePos?.let { tag.putIntArray("CorePos", intArrayOf(it.x, it.y, it.z)) }
    }

    override fun loadAdditional(tag: ValueInput) {
        super.loadAdditional(tag)
        tag.getIntArray("CorePos").ifPresent { a -> if (a.size >= 3) corePos = BlockPos(a[0], a[1], a[2]) }
    }

    override fun getUpdateTag(registries: HolderLookup.Provider): CompoundTag = saveWithoutMetadata(registries)
    override fun getUpdatePacket(): ClientboundBlockEntityDataPacket = ClientboundBlockEntityDataPacket.create(this)
}
